File: alarmclock/src/main/java/com/externalalarmclock/alarmclock/AlarmStore.java
package com.externalalarmclock.alarmclock;

import java.time.ZonedDateTime;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import com.externalalarmclock.lib.rpiws281x.RpiWs281xChannel;

public class AlarmStore {
	private Map<RpiWs281xChannel, ZonedDateTime> alarmTime = new HashMap<>();

	public ZonedDateTime getNextAlarm(RpiWs281xChannel channel) {
		return alarmTime.get(channel);
	}

	public void setNextAlarm(ZonedDateTime alarmTime) {
		for (RpiWs281xChannel channel : this.alarmTime.keySet()) {
			this.alarmTime.put(channel, alarmTime);
		}
	}

	public void addChannels(Collection<RpiWs281xChannel> channels) {
		channels.forEach(c -> alarmTime.put(c, null));
	}
}

File: alarmclock/src/main/kotlin/com/externalalarmclock/alarmclock/AlarmStore.kt
package com.externalalarmclock.alarmclock

import java.time.ZonedDateTime
import java.util.HashMap

import com.externalalarmclock.lib.rpiws281x.RpiWs281xChannel
import java.time.Duration

class AlarmStore {
    lateinit var wakeUpLightDuration: Duration
    private val alarmTime = HashMap<RpiWs281xChannel, Pair<ZonedDateTime?, ZonedDateTime?>?>()

    fun getNextAlarm(channel: RpiWs281xChannel): Pair<ZonedDateTime?, ZonedDateTime?>? {
        return alarmTime[channel]
    }

    fun setNextAlarm(alarmTime: ZonedDateTime?) {
        for (channel in this.alarmTime.keys) {
            this.alarmTime[channel] = alarmTime to alarmTime?.minus(wakeUpLightDuration)
        }
    }

    fun addChannels(channels: Collection<RpiWs281xChannel>) {
        channels.forEach { c -> alarmTime[c] = null }
    }
}
